r/R/transforms.R

```diff
@@ -812,3 +812,84 @@ multiply_cols <- function(mat, vec) {
   assert_true(length(vec) == ncol(mat))
   t(t(mat) * vec)
 }
+
+
+
+#################
+# Linear residual
+#################
+
+setClass("TransformLinearResidual", contains = "TransformedMatrix")
+setMethod("iterate_matrix", "TransformLinearResidual", function(x) {
+  res <- iterate_matrix(x@matrix)
+  
+  if (any(nrow(x@row_params) == 0, nrow(x@col_params) == 0)) {
+    return(res)
+  }
+  iterate_matrix_linear_residual_cpp(res, x@row_params, x@col_params)
+})
+setMethod("short_description", "TransformLinearResidual", function(x) {
+  # Return multiple lines, one for each transform active
+  res <- short_description(x@matrix)
+  if (any(nrow(x@row_params) == 0, nrow(x@col_params) == 0)) {
+    res <- c(res, sprintf("Linear regress out 0 variable"))
+    return(res)
+  }
+  res <- c(res, sprintf("Linear regress out %d variable(s)", nrow(x@row_params) - 1))
+  res
+})
+
+#' Linear Regression
+#'
+#' Calculate the residuals from a linear model fit. 
+#'
+#' @param mat An IterableMatrix
+#' @param latent_data Extra data to regress out, should be a `data.frame` where each column is a 
+#' variable to regress out.
+#' @param prediction_axis Where the predicted values are. Options include "row" (default) and "col". 
+#' The row number of `latent_data` should be equal to number of the other axis than `prediction_axis` 
+#' in `mat`.
+#' @return An IterableMatrix. If input an empty `latent_data`, will return the `mat` itself.
+#' @export
+regress_out <- function(mat, latent_data = NULL, prediction_axis = c("row", "col")) {
+  if (length(latent_data) == 0) {
+    return(mat)
+  }
```

Comment:
I don't think we should silently do nothing in this case. Either we should take away the `latent_data = NULL` default and have the function error out with no input variables, or we should make the function regress out an intercept-only model (possibly with an additional warning printed) when 0 variables or NULL are passed in.

e.g. something like this (though it would have to come after the `match.arg(prediction_axis)` call and we'd have to add an option for `latent_data` to be a matrix as suggested below)
```suggestion
  if (length(latent_data) == 0 || is.null(latent_data)) {
    prediction_dim <- if(prediction_axis == "row") nrow(mat) else ncol(mat)
    latent_data <- matrix(rep.int(1, prediction_dim), dimnames=list(NULL, "(Intercept)"))
  }
```